### ISSUE TITLE
Copying files from File Service to Blob Storage is extremely slow

### ISSUE THREAD

**Issue by arneeiri** [Azure/azure-storage-net-data-movement#105]
I copy around 7500 small mp3 files from a File Service share to a Blob Storage Container. Yesterday this operation took 5 hours.

Running the same operation from AzCopy takes around 2 minutes.

I have been in touch with Azure Support team and they have seen in their logs a large amount of 404 errors in their logs for blob resources that are part of this operation. I have not made any requests that should cause these 404 errors myself. I believe some part of the copy operation causes it.

I believe that this issue depends on the storage accounts as I am able to perform copy in other storage accounts more efficient. If you need more information about the storage accounts I use, please post an email address and I will send you more information.

I run this as part of an Azure Webjob. This is the code:

```
 public static class FileCopier
    {
        public static async Task CopyAudioFilesToPublicationFolder(CloudFileShare cloudFileShare, string sourceFolderName, CloudBlobDirectory destinationBlobDirectory)
        {
            var fileShare = cloudFileShare;
            var rootDirectory = fileShare.GetRootDirectoryReference();
            var directory = rootDirectory.GetDirectoryReference(sourceFolderName);
            
            TransferManager.Configurations.ParallelOperations = 64;
            ServicePointManager.DefaultConnectionLimit = Environment.ProcessorCount * 8;
            ServicePointManager.Expect100Continue = false;
            var stopwatch = Stopwatch.StartNew();

            await TransferManager.CopyDirectoryAsync(directory, destinationBlobDirectory, true, new CopyDirectoryOptions {Recursive = true}, new DirectoryTransferContext() {SetAttributesCallback = SetAttributes});
            stopwatch.Stop();
            Console.WriteLine($"{DateTime.Now.ToLongTimeString()}: Copy files took {stopwatch.ElapsedMilliseconds}ms");
        }

        private static void SetAttributes(object destination)
        {
            var cloudBlockBlob = (CloudBlockBlob) destination;
            cloudBlockBlob.Properties.CacheControl = "public, max-age=31536000";
        }
    }
```

**Comment by blueww** [comment]
@arneeiri 
Would you please share the AzCopy Command line? And do you run Azcopy and DMlib on same machine with same account? Then we can see why DMlib is so much slower than AzCopy.

Besides that, from you code, I see DMlib set isServiceCopy = true in below code, so it will use Server side copy (which don't have SLA, and performance handle by server). 

`await TransferManager.CopyDirectoryAsync(directory, destinationBlobDirectory, true, new CopyDirectoryOptions {Recursive = true}, new DirectoryTransferContext() {SetAttributesCallback = SetAttributes});`
            
In you AzCopy command, do you specify "/SyncCopy"? If so, AzCopy will use SyncCopy (download to memory and upload), and it has different performance with Server side copy. If so, you should set isServiceCopy =false in above code.

My mail is weiwei@microsoft.com. You send me the detail test account , and running time if need more investigation.

**Comment by blueww** [comment]
Already discuss in mail, and suggest to use /SyncCopy.
No further feedback receive for 2 months. So close the issue.

**Comment by arneeiri** [comment]
That is alright. I have already fixed this issue in production with
SyncCopy!

**Comment by afkhan517** [comment]
can you please suggest me how to use SyncCopy or any other way to copy files within azurefileshare? because TransferManager.CopyDirectoryAsync is too slower when copying directory within same ShareReference.
FYI : Directory contains only files(.pdf,.txt) rather than sub-directories in it.
Please help asap

**Comment by blueww** [comment]
@afkhan517 

You can set parameter isServiceCopy as false when run TransferManager.CopyDirectoryAsync(), then it will run SyncCopy (download & upload), instead of AsyncCopy (Server side Copy ).

The "Async" in the function name TransferManager.CopyDirectoryAsync(), means the function is Async and will return a Async task. The sync or Async copy is controlled by the parameter "isServiceCopy".

To run SyncCopy, it's better to using a big VM in same region with same location of the destination Storage Account, for better performance.

**Comment by afkhan517** [comment]
Actually the size of my source directory is 200 mb and making isServiceCopy to "false" will first download and then upload which again taking too long time as compared to isServiceCopy "true" and sometimes internet speed is also too weak and downloading and uploading 200 mb file takes long time. 
Currently copying directory using azurefileshare takes 5.23 min and copying the same using normal file copy(File.Copy) takes 3 min.
It would be a great help if you suggest something related to isServiceCopy "true" or any other way.
Thanks in advance.

**Comment by blueww** [comment]
For isServiceCopy "true", actually the copy process is handle by server, so we can't speed up it from client side. If your files count are big, set bigger `ServicePointManager.DefaultConnectionLimit` (default id CPU core * 8) might can make it faster since more files will copy in parallel.

Besides that, if your machine which runs Azcopy has low network bandwidth to Azure Server, SyncCopy might be slower since it need download to local memory and then upload. So we would suggest to run SyncCopy on a VM in same region with the destination Storage account.

**Comment by afkhan517** [comment]
@blueww 
"SyncCopy on a VM in same region with the destination Storage account"- we have it in same location
"If your files count are big, set bigger ServicePointManager.DefaultConnectionLimit (default id CPU core * 8) " - done with this but result is same.
And the strange thing is if i am iterating over FileInfo and using File.Copy then its taking the half time as compared to azurefileshare.
FYI : in source directory i have 4000 files and size of it is 200 mb
Please help asap.
Thanks in advance

**Comment by blueww** [comment]
@afkhan517 

Would you please check if you have set the 2 value in your process before any http request is sent to server:
```
ServicePointManager.DefaultConnectionLimit = Environment.ProcessorCount * 8; #or larger number
ServicePointManager.Expect100Continue = false;
```

And for " iterating over FileInfo and using File.Copy", do you mean you mount Azure File share to local then Copy it in local mounted share, or something else?

For the VM running AzCopy, what size is the VM?

**Comment by afkhan517** [comment]
@blueww 
if i am using this its takes upto 3 min -->
DirectoryInfo loSource = new DirectoryInfo(fsOldDirectoryPath);
                FileInfo[] files = loSource.GetFiles();
                var stopwatch = Stopwatch.StartNew();
                foreach (FileInfo file in files)
                {
                    file.CopyTo(Path.Combine(fsNewDirectoryPath, file.Name));
                }
                stopwatch.Stop();
                Console.WriteLine($"{DateTime.Now.ToLongTimeString()}: Copy files took {stopwatch.ElapsedMilliseconds}ms");

And for this its taking 5-6 min -->
 TransferManager.CopyDirectoryAsync(foCloudFileDirectorySource, foCloudFileDirectoryDestination,
                        true/* takes 5-6 min if true and 8-9 min if false*/,
                        new CopyDirectoryOptions()
                        {
                            Recursive = true,
                        },
                        new DirectoryTransferContext()).Wait();
And also have set this at top before any http request-->
ServicePointManager.DefaultConnectionLimit = Environment.ProcessorCount * 8;
ServicePointManager.Expect100Continue = false;

Total files : 4000
Size : 200 mb

Thanks in advance

**Comment by blueww** [comment]
For the above result, I assume you have mounted the File share to local, and use File.Copy(). Please confirm.

And would you like to try copying the File with AzCopy, and see if it have same performance with DMlib.
https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azcopy?toc=%2fazure%2fstorage%2fblobs%2ftoc.json 

The command should looks like：

> AzCopy /Source:https://myaccount1.file.core.windows.net/myshare1/dir1 /Dest:https://myaccount2.file.core.windows.net/myshare2/dir2 /SourceKey:key1 /DestKey:key2 /s /y 

for SyncCopy, add "/SyncCopy"
for verbose log, ass "/v"

If AzCopy and DMlib have difference performance, it should be cause by some issue in your code. 
If AzCopy and DMlib have same performance, the issue should be related with the environment. Would you please share more details about your environment, like the VM size/location, the test account location.

For your information, we have tested AzureFile->AzureFile Sync Copy on a D14 VM, the data size is  4096*512KB (total 2GB), and it use about 50s to finish the SyncCopy.

**Comment by afkhan517** [comment]
@blueww 
Thanks a lot dear. AzCopy Works for me.
just one doubt how to exclude sub-directory using azcopy cmd and also pattern for all files?
what if i don't want to copy sub-directories from source and just want to copy only files from it.
Thanks in advance

**Comment by blueww** [comment]
@afkhan517 

If AzCopy has difference performance with DMlib, then it must be some code issue when using DMlib (since Azcopy based on DMlib).

The most possible cause is, you must set the following at top of your process start, before any http request send from your process (even the request not from DMlib).

ServicePointManager.DefaultConnectionLimit = Environment.ProcessorCount * 8; #or larger number
ServicePointManager.Expect100Continue = false;

**Comment by afkhan517** [comment]
Hi @blueww 

I have 4200 files in my source directory and copying with AzCopy it copies only 440 files and then return.

cmd : AzCopy /Source: /Dest: /SourceKey: /DestKey: /Z:  /SyncCopy /S /XO /XN

FYI : destination directory is always empty as we always copy to "GUID" string folder so there is no occurence of any files

It is just unexpected.

Please help asap.

**Comment by blueww** [comment]
@afkhan517 
Please open a new issue for this, as this is a different issue from before issue.

As I see, you have added /xo /xn, so the file will only be copy when the destination has no file with same name.

I don't quite understand "destination directory is always empty as we always copy to "GUID" string folder so there is no occurence of any files". 
You might can give some sample. and the source and dest link sample. (hide any credential)

Anyway, it's most probably that some files not copy since the destination already has file with same name, and you use /xo /xn parameter.

**Comment by varun867** [comment]
Does the TransferManager.CopyDirectoryAsync still work for copying files from fileshare to a Blob container?  It worked from Container to container, but i cannot get it to work when source is fileshare.
the method warns me that it is obsolete, but able to copy  from Container to container, I need help when source is a mounted File share.
`public  async Task CopyFilesToBlobContainer (CloudFileShare cloudFileShare, string sourceFolderName, CloudBlobDirectory destinationBlobDirectory)
        {
            var fileShare = cloudFileShare;
            var rootDirectory = fileShare.GetRootDirectoryReference();
            var directory = rootDirectory.GetDirectoryReference(sourceFolderName);

            TransferManager.Configurations.ParallelOperations = 64;
            System.Net.ServicePointManager.DefaultConnectionLimit = Environment.ProcessorCount * 8;
            System.Net.ServicePointManager.Expect100Continue = false;
            var stopwatch = Stopwatch.StartNew();

            DirectoryTransferContext context = new DirectoryTransferContext();

            context.SetAttributesCallbackAsync = async (destination) =>
            {
                var cloudBlockBlob = (CloudBlockBlob)destination;
                cloudBlockBlob.Properties.CacheControl = "public, max-age=31536000";
            };
            context.ShouldOverwriteCallbackAsync = TransferContext.ForceOverwrite;

            var trasferStatus  = await TransferManager.CopyDirectoryAsync(directory, destinationBlobDirectory, true, new CopyDirectoryOptions { Recursive = true }, context); //,new DirectoryTransferContext() { SetAttributesCallback = SetAttributes })
            stopwatch.Stop();
            Console.WriteLine($"{DateTime.Now.ToLongTimeString()}: Copy files took {stopwatch.ElapsedMilliseconds}ms");

        }`
What am i missing? Since this method is Onbolete, Is there a better way of moving directories?

**Comment by blueww** [comment]
@varun867 
DMlib has a new way to input copy method. Before there are only 2 choice: SyncCopy  and ServiceSideAsyncCopy, so we use a bool to represent it. Now there are 3 choice.
```
 public enum CopyMethod
    {
        SyncCopy = 0,
        ServiceSideAsyncCopy = 1,
        ServiceSideSyncCopy = 2
    }
```
As you choose isServiceCopy = true before, so you can change or original code :
` var trasferStatus = await TransferManager.CopyDirectoryAsync(directory, destinationBlobDirectory, true, new CopyDirectoryOptions { Recursive = true }, context);`
to 
` var trasferStatus = await TransferManager.CopyDirectoryAsync(directory, destinationBlobDirectory, CopyMethod.ServiceSideAsyncCopy, new CopyDirectoryOptions { Recursive = true }, context);`

Anyway, you can try other 2 options.
